Support adding services while `ServiceGroup` is running

sliemeobn opened the pull request:
somewhat straight-forward implementation draft of 
https://github.com/swift-server/swift-service-lifecycle/issues/185

choices were made, what do we think?

FranzBusch commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -106,6 +108,37 @@ public actor ServiceGroup: Sendable, Service {
+            await addedServiceChannel.send(serviceConfiguration)
+
+        case .finished:
+            return
```
I think this case definitely needs to throw an error to indicate something went wrong

FranzBusch replied:
I think that's fair. Let's ignore my comment then. `addService` is a best effort thing and if the group is finished or shutting down it won't run the new service. I think we just need to add a debug log for when that happens.

FranzBusch commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -310,6 +333,49 @@ public actor ServiceGroup: Sendable, Service {
+                            index: services.count - 1
+                        )
+                    }
+                } onCancel: {
+                    addedServiceChannel.finish()
+                }
```
This shouldn't be necessary since the channel setups up cancellation handlers as well and the for loop above will finish.

FranzBusch replied:
If that's the case it sounds like a bug in the `AsyncChannel` to me

FranzBusch commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -310,6 +333,49 @@ public actor ServiceGroup: Sendable, Service {
+                        )
+
+                        let gracefulShutdownManager = GracefulShutdownManager()
+                        gracefulShutdownManagers.append(gracefulShutdownManager)
+                        services.append(serviceConfiguration)
```
I don't see how this is safe from a concurrency point of view since the services & gracefulShutdownManagers variables are in a different task. This is probably causing concurrency issues in Swift 6.

sliemeobn replied:
I am not sure I follow. isn't everything actor-isolated to the `ServiceGroup`?

FranzBusch replied:
The child task isn't actor isolated since the closures of `addTask` are `sending`. Maybe I'm wrong but this looks to me like something that would result in an error in language mode 6. Could you check that real quick?

sliemeobn replied:
you are absolutely correct (I am not used to writing code without strict concurrency checking any more ; )
ok then, this settles the question about the escaping task group as well - we should funnel everything to the main loop on the actor somehow...

FranzBusch commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -310,6 +333,49 @@ public actor ServiceGroup: Sendable, Service {
                 "We did not create a graceful shutdown manager per service"
             )
 
+            var _unownedTaskGroupHandledCarefully = group
```
I think we need to come up with a different solution here than unsafe escaping the task group into a child task.

sliemeobn replied:
^^ I figured you'd say that.
I am pretty sure this is perfectly safe as written because we do it "withoutActuallyEscaping" if you will.

there is no good alternative I can think of would be not be either a convoluted merging of the two streams (ie: the group itself and the added services channel - which would then add to the group?) or a complete overhaul of the run loop and wind-down loop.

FranzBusch replied:
> I am pretty sure this is perfectly safe as written because we do it "withoutActuallyEscaping" if you will.
I agree that this is most likely safe but we can't prove it to the compiler sadly. Let me think a bit about it.

FranzBusch replied:
Just after having written this I actually thought of a something where this isn't safe. The underlying task group implementation does in fact no guarantee that you can call `addTask` concurrently nor does the `group.next()` support concurrent iteration. The fact that this works is pure luck and nothing that we can expect to stay this way.

sliemeobn replied:
yeah, as mentioned above: after realizing that `.addTask` does not run on the actors isolation the whole thing falls apart anyway... we'll need a merged async sequence I think...

FranzBusch reviewed (commented):
This is a great start. Thanks for working on this. I left some feedback inline.

sliemeobn commented:
^^ I was just coding this exact thing, but I wasn't sure if it is too crazy ; )
I'll see how it goes...

sliemeobn commented:
@FranzBusch I managed to clean up the channel reading, I think that part is good enough now.

The remaining question is what to do with added services during shutdown or cancel?

If we want a reliable back-propagation of "has it run or not?" we'll need to pass along a continuation I think and we'd need to make sure to fully drain and actively "reject" all added service until we set the status to `.finished`.

A lot of brittle and racy signaling... if we simply ignored (like now) life would be a lot easier ; )

FranzBusch commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -779,6 +847,43 @@ public actor ServiceGroup: Sendable, Service {
     }
 }
 
+extension ThrowingTaskGroup where Failure == Error, ChildTaskResult == ServiceGroup.ChildTaskResult {
```
This is more of a nit but I would prefer if we don't spell these methods as extensions on the group. I would prefer if we just make them private methods on `ServiceGroup` itself.

sliemeobn replied:
done

FranzBusch commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -779,6 +847,43 @@ public actor ServiceGroup: Sendable, Service {
+
+    mutating func addAddedServiceListenerTask(_ channel: AsyncChannel<ServiceGroupConfiguration.ServiceConfiguration>) {
+        self.addTask {
+            return await withTaskCancellationHandler {
```
This still seems wrong. We need to figure out why we need this. `AsyncChannel` does handle cancellation itself normally.

sliemeobn replied:
It might be as designed though. especially with out multi-iterator `next`ing off one at a time setup, how else is the channel supposed to know that it is no longer listened to?

FranzBusch replied:
Oh thanks for this comment. Yes I remember now that async channel ignores cancellation on both sides since it is a multi producer multi consumer channel.

FranzBusch commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -717,6 +780,11 @@ public actor ServiceGroup: Sendable, Service {
                     break
+
+                case .newServiceAdded:
+                    // TBD: How do we treat added services during graceful shutdown?
+                    // Currently, we ignore them - but we make sure that `run` is never called
+                    break
```
I think that's fine. If we are shutting down we can just ignore them.

FranzBusch commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -310,12 +334,48 @@ public actor ServiceGroup: Sendable, Service {
                 "We did not create a graceful shutdown manager per service"
             )
 
+            group.addTask {
+                // This child task is waiting forever until the group gets cancelled.
+                let (stream, _) = AsyncStream.makeStream(of: Void.self)
+                await stream.first { _ in true }
+                return .cancellationCaught
+            }
+
+            // Adds a task that listens to added services and funnels them into the task group
+            group.addAddedServiceListenerTask(addedServiceChannel)
+
             // We are going to wait for any of the services to finish or
             // the signal sequence to throw an error.
             while !group.isEmpty {
-                let result: ChildTaskResult? = try await group.next()
+                let nextEvent = try await group.next()
```
Any reason why we need to rename this?

sliemeobn replied:
nope, none at all

sliemeobn commented:
thanks @FranzBusch for the quick reviews!

I addressed the feedback from above, but I am still a bit scared of the "best effort" nature of it.

So, as a suggestion, I renamed the function to `addServiceUnlessShutdown` to better communicate its best-effort nature (similar to `TaskGroup.addTaskUnlessCanceled`).

This would also allow us to one day add a `addService` that maybe always runs and immediately shuts down, and maybe throws if the group was canceled or has shutdown. (or similar, should the need arise)

wdyt?

FranzBusch commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -777,6 +850,45 @@ public actor ServiceGroup: Sendable, Service {
+
+                return .gracefulShutdownFinished
+            } onCancel: {
+                // Without this we can get stuck in `addService` if the group
```
Can we change this comment to reflect our discussion above and just say `AsyncChannel` is MPSC hence we need to shut it down on cancel.

FranzBusch reviewed (commented):
This looks great. I'm happy with the new name and the implementation looks good to me. Just one last nit around the comment

FranzBusch commented:
@sliemeobn can you fix the license header check please

sliemeobn commented:
> can you fix the license header check please
@FranzBusch oops, done

(is there an easy way to run the soundness checks locally?)

FranzBusch commented:
> > can you fix the license header check please
> 
> @FranzBusch oops, done
> 
> (is there an easy way to run the soundness checks locally?)

Yeah you can use `act` to run all of it locally. There is some documentation here: https://github.com/swiftlang/github-workflows?tab=readme-ov-file#running-workflows-locally

czechboy0 commented on Sources/ServiceLifecycle/ServiceGroup.swift:
```diff
@@ -106,6 +108,38 @@ public actor ServiceGroup: Sendable, Service {
+    ///
+    /// If the group is currently running, the added service will be started immediately.
+    /// If the group is gracefully shutting down, cancelling, or already finished, the added service will not be started.
+    /// - Parameters:
```
Apologies for the late review - it'd be great to call out that this can also be used for adding services before the group was started. Right now the comment leaves that ambiguous of what happens.
